service/frontend/workflow_handler.go

```
@@ -5261,12 +5261,12 @@ func (wh *WorkflowHandler) validateCallbackURL(ns namespace.Name, rawURL string)
 	if err != nil {
 		return err
 	}
-	if !(u.Scheme == "http" || u.Scheme == "https") {
+	if !(u.Scheme == "http" || u.Scheme == "https" || u.Scheme == "temporal") {
 		return status.Errorf(codes.InvalidArgument, "invalid url: unknown scheme: %v", u)
 	}
 	for _, cfg := range wh.config.CallbackEndpointConfigs(ns.String()) {
-		if cfg.Regexp.MatchString(u.Host) {
-			if u.Scheme == "http" && !cfg.AllowInsecure {
+		if cfg.MatchHost(u.Host) {
+			if !cfg.SchemeAllowed(u.Scheme) {
```

I don't think you need this check since you are already checking the known schemes above on L5264

---

SchemeAllowed also checks insecure vs not secure. it's configured on a per endpoint basis. I'd prefer to return an error from here so I can return an error for if the scheme is not support outright or if insecure is just not allowed for the given match. I could just make another method that check if the schema itself is valid.